### ISSUE TITLE
FreeBSD: fail2ban should not stick forever on ipfw entries deleted by hand

### ISSUE THREAD

**Issue by ache** [fail2ban/fail2ban#893]
Currently if some entry is deleted by ipfw manually, it stick forever in sqlite database and fail2ban attempts to delete it over and over. Fail2ban simple should understand this ipfw error message (ESRCH - "No such process") as _success_ to prevent it:
014-12-22 20:22:46,060 fail2ban.action         [57792]: ERROR   ipfw table 2 delete 117.21.225.176 -- stderr: 'ipfw: setsockopt(IP_FW_TABLE_XDEL): No such process\n'

**Comment by ache** [comment]
For address adding you also need to handle this error as success:
ERROR   ipfw table 2 add 62.210.178.242 -- stderr: 'ipfw: setsockopt(IP_FW_TABLE_XADD): File exists\n'

**Comment by leeclemens** [comment]
What if you use fail2ban-client to unban it, rather than modifying ipfw directly (which fail2ban won't know about)?

**Comment by ache** [comment]
This is not limited to direct modification. Another example is restart after machine crash (ipfw tables are empty, but fail2ban will try forever from this moment to unban already unbanned). Another example is fail2ban crash (I see it few times in python sqlite3 module). Etc. Moreover, with each such case the number of unneeded attempts and junk log entries increases over time (since fail2ban don't know a way to decrease it).

The general rule for all complex cases is treating attempt to delete of already deleted as OK and attempt to add already present as OK too.

**Comment by leeclemens** [comment]
My response was based upon your previous comments. Does fail2ban-client unban resolve this particular issue? (I realize you may see other issues, in addition to what you originally raised here.)

**Comment by ache** [comment]
No, both fail2ban client and daemon fails with errors I already quoted exactly in the start of thread.
First error is delete already deleted. Second error is add already added.
This two should be no errors at all!

Nothing from fail2ban side can drop this stale entries afterwards (excepting external removing perfectly valid fail2ban database and rebooting the machine) and its number increased with each incident.

**Comment by ache** [comment]
To fix that you need to check not only ipfw exit code but the error string
'ipfw: setsockopt(IP_FW_TABLE_XDEL): No such process\n'
on deleting and the error string
'ipfw: setsockopt(IP_FW_TABLE_XADD): File exists\n'
on adding.

**Comment by leeclemens** [comment]
Care to submit a PR?

**Comment by ache** [comment]
Does fail2ban have PRs system? I don't now how and where.

**Comment by leeclemens** [comment]
Yes, it is hosted on github so you can clone the repo, make modifications and submit a Pull Request.

**Comment by ache** [comment]
Sorry, I treat your PR as Problem Report )) and I am familiar with github in view or commenting modes only. It seems some shell code needs to be added to 'actionban' and 'actionunban'. Something like that (untested):
```
actionban = ipfw table <table> add <ip> 2> /tmp/fail2ban$$ || [ `cat /tmp/fail2ban$$` = 'ipfw: setsockopt(IP_FW_TABLE_XADD): File exists' ]

actionunban = ipfw table <table> delete <ip> 2> /tmp/fail2ban$$ || [ `cat /tmp/fail2ban$$` = 'ipfw: setsockopt(IP_FW_TABLE_XDEL): No such process' ]
```
Temp file name makes me nervous, it is insecure with $$. It will be better to use
```
TMPFILE=`mktemp /tmp/fail2ban.XXXXXX` && ...
```
instead, prepending at the start of each command.

**Comment by leeclemens** [comment]
I don't currently have a box using ipfw to test with, I'd suggest creating a Pull Request for someone to see and incorporate your changes.

**Comment by ache** [comment]
This code is what first comes to mind just to illustrate idea, it is not for real merging. Temp file usage is unneeded, stderr can be assigned to shell variable. I'll come with optimized and tested code later.

**Comment by leeclemens** [comment]
Thank you for understanding, any helpful contribution is always welcome!

**Comment by ache** [comment]
Now tested and optimized. Change in bsd-ipfw.conf:
```
actionban = e=`ipfw table <table> add <ip> 2>&1`; x=$?; [ $x -eq 0 -o "$e" = 'ipfw: setsockopt(IP_FW_TABLE_XADD): File exists' ] || { echo "$e" 1>&2; exit $x; }

actionunban = e=`ipfw table <table> delete <ip> 2>&1`; x=$?; [ $x -eq 0 -o "$e" = 'ipfw: setsockopt(IP_FW_TABLE_XDEL): No such process' ] || { echo "$e" 1>&2; exit $x; }
```

**Comment by pgib** [comment]
:+1: Thank you for this.

**Comment by yarikoptic** [comment]
was closed via #952
